qiskit_machine_learning/neural_networks/effective_dimension.py

```diff
@@ -0,0 +1,349 @@
+#
+# (C) Copyright IBM 2022.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""An implementation of the effective dimension algorithm."""
+
+import time
+from typing import Optional, Union, List, Callable, Tuple
+
+import numpy as np
+from scipy.special import logsumexp
+
+from qiskit.utils import algorithm_globals
+
+from .opflow_qnn import OpflowQNN
+from .neural_network import NeuralNetwork
+
+
+class EffectiveDimension:
+    """
+    This class computes the global effective dimension for Qiskit
+    :class:`~qiskit_machine_learning.neural_networks.NeuralNetwork`s
+    following the definition used in [1].
+
+    **References**
+        [1]: Abbas et al., The power of quantum neural networks. `The power of QNNs
+            <https://arxiv.org/pdf/2011.00027.pdf>`_
+    """
+
+    def __init__(
+        self,
+        qnn: NeuralNetwork,
+        params: Optional[Union[List[float], np.ndarray, int]] = 1,
+        inputs: Optional[Union[List[float], np.ndarray, int]] = 1,
+        callback: Optional[Callable[[str], None]] = None,
+    ) -> None:
+
+        """
+        Args:
+            qnn: A Qiskit :class:`~qiskit_machine_learning.neural_networks.NeuralNetwork`,
+                with a specific dimension ``(num_weights)`` that will determine the shape of the
+                Fisher Information Matrix ``(num_inputs * num_params, num_weights, num_weights)``
+                used to compute the global effective dimension for a set of ``inputs``,
+                of shape ``(num_inputs, qnn_input_size)``,
+                and ``params``, of shape ``(num_params, num_weights)``.
+            params: An array of neural network parameters (weights), of shape
+                ``(num_params, num_weights)``, or an ``int`` to indicate the number of parameter
+                sets to sample randomly from a uniform distribution. By default, ``params = 1``.
+            inputs: An array of inputs to the neural network, of shape
+                ``(num_inputs, qnn_input_size)``, or an ``int`` to indicate the number of input
+                sets to sample randomly from a normal distribution. By default, ``inputs = 1``.
+            callback: A callback function for the Monte Carlo sampling.
+        """
+
+        # Store arguments
+        self._params = None
+        self._inputs = None
+        self._num_params = 1
+        self._num_inputs = 1
+        self._model = qnn
+        self._callback = callback
+
+        # Define inputs and parameters
+        self.params = params  # type: ignore
+        # input setter uses self._model
+        self.inputs = inputs  # type: ignore
```

Comment:
Is the type ignore here needed? From what I can see the `params` type in the constructor matches the setter (which it should)